csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -197,10 +316,63 @@ public CompressionResult CompressPayload()
             _compressionMinSize = compressionMinSize;
         }
 
-        internal void Finish(OutputStream.Position encapsulationEnd)
+        private OutputStream WriteBinaryContextEntryKey(int key)
         {
-            Size = Payload.GetByteCount();
+            if (Protocol == Protocol.Ice1)
+            {
+                throw new NotSupportedException("binary context is not supported with Ice1 protocol");
+            }
+
+            if (IsSealed)
+            {
+                throw new InvalidOperationException("cannot modify a sealed frame");
+            }
+
+            OutputStream.Position encapsulationEnd = _encapsulationEnd ??
```

Comment:
It's unclear to me when this could happen. We're not always filling the payload while we construct the OutgoingFrame?

Reply:
`WriteBinaryContextEntryKey` is called from each `AddBinaryContextEntry` validates the key, and this checks ensure that is not legal to add binary context entries before you finish writing the encaps, because everything is wrote to the same buffer `List<ArraySegment<byte>> Payload` , if we write the binary context to its own buffer that wouldn't be a problem

Reply:
Replaced by an assert